core/renderers/common/drawer.js

```diff
@@ -149,7 +150,8 @@ class Drawer {
         this.outlinePath_ += this.constants_.OUTSIDE_CORNERS.topLeft;
       } else if (Types.isRightRoundedCorner(elem)) {
         this.outlinePath_ += this.constants_.OUTSIDE_CORNERS.topRight;
-      } else if (Types.isPreviousConnection(elem)) {
+      } else if (
+          Types.isPreviousConnection(elem) && elem instanceof Connection) {
```

Comment:
I'm assuming that `Types.isPreviousConnection(elem)` implies that `elem instanceof Connection` and you're just helping the compiler figure that out, but if that's not so beware that this could result in a change of behaviour, where cases that would previously have matched this `if` clause could end up falling through and potentially matching later ones.

A slightly more conservative approach to adding a type check here would be:
```JS
      } else if (Types.isPreviousConnection(elem)) {
        if(!(elem instanceof Connection)) throw new TypeError(/*...*/);
```

(Here and various other places.)

Reply:
I'm pretty sure that these `instanceof` checks shouldn't modify behavior, especially because we rely on properties of the checked class existing.

Honestly, I'm not sure why `Types` exists at all. It looks like it was [meant to unify a bunch of ad-hoc "type" checks](https://github.com/google/blockly/pull/2950/files). But it seems like `instanceof` works just as well.

Reply:
Just bumping this to check your opinion before merging @cpcallen

Reply:
LGTM.